hdt-qs-backend/src/main/java/com/the_qa_company/qendpoint/federation/WikibaseLabelService.java

```diff
@@ -0,0 +1,172 @@
+		System.out.println("Should never pass here, report the query if this is the case!");
+		return false;
+	}
+
+	@Override
+	public CloseableIteration<BindingSet, QueryEvaluationException> select(Service service, Set<String> set,
+			BindingSet bindingSet, String s) throws QueryEvaluationException {
+		return null;
+	}
+
+	@Override
+	public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(Service service,
+			CloseableIteration<BindingSet, QueryEvaluationException> closeableIteration, String s)
+			throws QueryEvaluationException {
+		TupleExpr tupleExpr = service.getArg();
+		// currently implements only the automatic mode
+		// https://en.wikibooks.org/wiki/SPARQL/SERVICE_-_Label
+		// the SERVICE clause contains only one triple pattern that specifies
+		// the language
+		if (tupleExpr instanceof StatementPattern) {
+			StatementPattern statement = (StatementPattern) tupleExpr;
+			// the predicate must be <http://wikiba.se/ontology#language>
+			if (statement.getPredicateVar().getValue().stringValue().equals("http://wikiba.se/ontology#language")) {
+				// the object must be a literal without language tag and without
+				// datatype
+				if (statement.getObjectVar().getValue().isLiteral()) {
+					Literal literal = (Literal) statement.getObjectVar().getValue();
+					if (!literal.getLanguage().isPresent() && literal.getDatatype() == XSD.STRING) {
+						List<String> languages = Arrays.asList(literal.getLabel().split(","));
+						if (languages.size() > 0) {
+							return new CloseableIteration<BindingSet, QueryEvaluationException>() {
+								@Override
+								public void close() throws QueryEvaluationException {
+									closeableIteration.close();
+								}
+
+								@Override
+								public boolean hasNext() throws QueryEvaluationException {
+									return closeableIteration.hasNext();
+								}
+
+								@Override
+								public BindingSet next() throws QueryEvaluationException {
+									return expandBindingSet(closeableIteration.next(), languages);
+								}
+
+								@Override
+								public void remove() throws QueryEvaluationException {
+									closeableIteration.remove();
+								}
+							};
+						}
+					}
+
+				}
+			}
+		}
+		return null;
+	}
+
+	private BindingSet expandBindingSet(BindingSet bindingSet, List<String> languages) {
+		ArrayList<String> namesWithLabels = new ArrayList<>();
+		ArrayList<Value> valuesWithLabels = new ArrayList<>();
+		// according to https://en.wikibooks.org/wiki/SPARQL/SERVICE_-_Label 3
+		// properties have to be expanded
+		IRI[] expantionProperties = new IRI[] { RDFS.LABEL, SKOS.ALT_LABEL,
+				vf.createIRI("https://schema.org/description") };
+		String[] expantionNameSuffix = new String[] { "Label", "AltLabel", "Description" };
+		for (int e = 0; e < 3; e++) {
+			List<String> languageLabels = new ArrayList<String>(Collections.nCopies(languages.size(), (String) null));
```

Comment:
Knowing we won't change the size of this array, a simple String array would be faster and less confusing I think?

Reply:
done